pkg/operator/ceph/disruption/clusterdisruption/osd.go

```diff
@@ -338,24 +338,9 @@ func (r *ReconcileClusterDisruption) reconcilePDBsForOSDs(
 	}
 
 	switch {
-	// osd is down but pgs are active+clean
-	case osdDown && pgClean:
```

Comment:
If the PGs are clean, the intent is that the PDBs will again reset after some timeout. Perhaps 30s was just too short. What about a timeout of 5 or 10 minutes? @sp98 thoughts?

Reply:
case osdDown && pgClean:

- This case would handle the scenario when OSD was down but PGs are still clean. 
- This can happen if OSD went down and it took some time for ceph to update the pg down status or for rook to read the pg down status. 
- So we wait for around 30 seconds after the OSD went down to confirm if the PGs went down or not. 
- If the PGs didn't go down after 30 seconds, we reset (#350) back to normal state and allow more OSDs to be drained.

Reply:
So the whole idea of the 30 second time period is to give enough time to Rook to read the `pg status` from correctly.

Reply:
If the OSD is down but PGs are acitive+clean, why would we want to even enable the blocking PDBs? Shouldn't we allow the next OSD to drain as data is safe?

Reply:
@sp98 If there are multiple OSDs on the node, the only way for a node to be drained is if the blocking PDBs are applied to other nodes or zones to allow this one to drain. If I'm understanding the timeout correctly, perhaps we need a longer timeout if there are multiple OSDs on a node, so the local node/zone can be allowed to drain.

Reply:
> So the whole idea of the 30 second time period is to give enough time to Rook to read the pg status from correctly.

I understood the waiting for some time is necessary after detecting a down osd.

>  If I'm understanding the timeout correctly, perhaps we need a longer timeout if there are multiple OSDs on a node, so the local node/zone can be allowed to drain.

I don't think lengthening timeout solves this problem. Could you elaborate it a little more? In my understanding, even if PDB configuration is reset, `maxUnavablable` is still 1 and any other OSDs in the same node can't become down.

Reply:
@ushitora-anqou Let me know your thoughts about my [comment](https://github.com/rook/rook/pull/13511#discussion_r1444592625) above

Reply:
(Trying to understand the scenario for which this PR is created)

Background: 
We create two types of pdbs:
- Main (default):
   → This has `maxUnavailable` set to 1.
   → So we always allow 1 osd to go down in case of drain event.
- Blocking:
   → These pdbs are created on the failure domains other than the one that is currently being drained. 
   → Here `maxUnavailable` is set to 0. 
   → So this pdb blocks any osd drain on these failure domains. 

Lets take an example scenario with `node` as the failure domain.

Node1
  osd.0
  osd.1
Node2
  osd.3
  osd.4
Node3
  osd.5
  osd.6

Node1 is drained
- lets say `osd.0` goes down first.
- `maxUnavailable` of the main(default) pdb becomes 0. So no more node drained allowed on this node. 
- Two scenarios can happen:

a) Scenario-1 : This happens most of the time. 
1.pgs become unheatlhy after draining of `osd.0`
2.After identifying that PGs are unhealthy and osd is down, Rook removes the main(default) pdb and creates blocking pdbs for `Node2` and `Node3` with `maxUnavailable=0`.  Now user can't drain `node2` and `node3`
3.remanining osds on `node1` can get drained now because the main(default) pdb was removed.

b) Scenario-2: (This PR is about this scenario)
1. pgs are still active+clean after draining of `osd.0`. ( Could happen if there is no data in that OSD. Any other reason?)
2. Rook identifies that OSD is down but pgs are clean. So rook does not create any blocking PDB and does not remove the main(default) PDB that now has `maxUnavailable=0` 
3. So drain of `osd.1` is blocked now. This is not what the user was expecting. User still wants osd.1 to be drained if pgs are active+clean.

Reply:
@ushitora-anqou are you trying to solve scenario-2 with this PR?

Reply:
@sp98 Yes, that's right.

Reply:
Thanks for confirming @ushitora-anqou 
I'll check the PR again in a day or two.

Reply:
@sp98 Hi. Could you please review this PR again?

Reply:
@ushitora-anqou did you get a chance to test this out locally? It should be tested for both the situations when PGs become unhealthy and when PGs remain active+clean, after the node drain.

Reply:
I tested this patch in my local (minikube) environment before submitting it here. I'll post the script to reproduce the result here soon.

Reply:
I re-confirmed that my PR works. I uploaded the scripts I used here: https://github.com/ushitora-anqou/rook-on-minikube/tree/rook-13511

Here's the result:

I deployed a Rook/Ceph cluster using my PR with 3 nodes and 6 OSDs, and also deployed a RGW.
```
$ kubectl exec -n rook-ceph -it deploy/rook-ceph-tools -- ceph osd tree
ID  CLASS  WEIGHT   TYPE NAME              STATUS  REWEIGHT  PRI-AFF
-1         0.00595  root default                                    
-7         0.00198      host minikube                               
 4    ssd  0.00099          osd.4              up   1.00000  1.00000
 5    ssd  0.00099          osd.5              up   1.00000  1.00000
-5         0.00198      host minikube-m02                           
 1    ssd  0.00099          osd.1              up   1.00000  1.00000
 3    ssd  0.00099          osd.3              up   1.00000  1.00000
-3         0.00198      host minikube-m03                           
 0    ssd  0.00099          osd.0              up   1.00000  1.00000
 2    ssd  0.00099          osd.2              up   1.00000  1.00000

$ kubectl exec -n rook-ceph -it deploy/rook-ceph-tools -- ceph osd df
ID  CLASS  WEIGHT   REWEIGHT  SIZE   RAW USE  DATA     OMAP  META     AVAIL     %USE  VAR   PGS  STATUS
 4    ssd  0.00099   1.00000  1 GiB  9.2 MiB  868 KiB   0 B  8.3 MiB  1015 MiB  0.90  0.83   46      up
 5    ssd  0.00099   1.00000  1 GiB  9.8 MiB  1.3 MiB   0 B  8.6 MiB  1014 MiB  0.96  0.90   43      up
 1    ssd  0.00099   1.00000  1 GiB   10 MiB  1.2 MiB   0 B  8.8 MiB  1014 MiB  0.98  0.91   31      up
 3    ssd  0.00099   1.00000  1 GiB   13 MiB  904 KiB   0 B   13 MiB  1011 MiB  1.31  1.22   58      up
 0    ssd  0.00099   1.00000  1 GiB   14 MiB  932 KiB   0 B   13 MiB  1010 MiB  1.38  1.28   50      up
 2    ssd  0.00099   1.00000  1 GiB  9.3 MiB  1.2 MiB   0 B  8.1 MiB  1015 MiB  0.91  0.85   39      up
                       TOTAL  6 GiB   66 MiB  6.4 MiB   0 B   60 MiB   5.9 GiB  1.07                   
MIN/MAX VAR: 0.83/1.28  STDDEV: 0.20
```

The default PDB was correctly set:
```
$ kubectl get pdb -n rook-ceph
NAME            MIN AVAILABLE   MAX UNAVAILABLE   ALLOWED DISRUPTIONS   AGE
rook-ceph-osd   N/A             1                 1                     106s
```

Then, I cordoned a node and deleted a OSD pod on the node to emulate draining of the node:
```
$ kubectl cordon minikube-m02
node/minikube-m02 cordoned

$ kubectl delete pod -n rook-ceph rook-ceph-osd-1-dfb8dfc98-chmbj
pod "rook-ceph-osd-1-dfb8dfc98-chmbj" deleted
```

Then, the blocking PDBs were successfully created (as before my PR is applied):
```
$ kubectl logs -n rook-ceph rook-ceph-operator-6644798f58-bj8t4 -f
2024-02-02 01:02:57.355897 I | clusterdisruption-controller: osd "rook-ceph-osd-1" is down and a possible node drain is detected
2024-02-02 01:02:57.788077 I | clusterdisruption-controller: osd is down in failure domain "minikube-m02". pg health: "cluster is not fully clean. PGs: [{StateName:active+clean Count:57} {StateName:active+undersized Count:22} {StateName:active+undersized+degraded Count:10}]"
2024-02-02 01:02:59.094658 I | clusterdisruption-controller: creating temporary blocking pdb "rook-ceph-osd-host-minikube" with maxUnavailable=0 for "host" failure domain "minikube"
2024-02-02 01:02:59.102011 I | clusterdisruption-controller: creating temporary blocking pdb "rook-ceph-osd-host-minikube-m03" with maxUnavailable=0 for "host" failure domain "minikube-m03"
2024-02-02 01:02:59.107224 I | clusterdisruption-controller: deleting the default pdb "rook-ceph-osd" with maxUnavailable=1 for all osd

$ kubectl get pdb -n rook-ceph ### <------------ The PDBs are correctly set
NAME                              MIN AVAILABLE   MAX UNAVAILABLE   ALLOWED DISRUPTIONS   AGE
rook-ceph-osd-host-minikube       N/A             0                 0                     2m14s
rook-ceph-osd-host-minikube-m03   N/A             0                 0                     2m14s
```

Next, I uncordoned the node, and made the OSD have no PGs:
```
$ kubectl uncordon minikube-m02

$ kubectl get pdb -n rook-ceph
NAME            MIN AVAILABLE   MAX UNAVAILABLE   ALLOWED DISRUPTIONS   AGE
rook-ceph-osd   N/A             1                 1                     5s

$ kubectl exec -n rook-ceph deploy/rook-ceph-tools -- ceph osd crush reweight osd.1 0
reweighted item id 1 name 'osd.1' to 0 in crush map

$ kubectl exec -n rook-ceph deploy/rook-ceph-tools -- ceph osd df
ID  CLASS  WEIGHT   REWEIGHT  SIZE   RAW USE  DATA     OMAP     META     AVAIL     %USE  VAR   PGS  STATUS
 4    ssd  0.00099   1.00000  1 GiB  9.7 MiB  1.0 MiB      0 B  8.7 MiB  1014 MiB  0.95  0.88   49      up
 5    ssd  0.00099   1.00000  1 GiB   10 MiB  1.4 MiB      0 B  8.9 MiB  1014 MiB  1.01  0.93   40      up
 1    ssd        0   1.00000  1 GiB  6.7 MiB  956 KiB    1 KiB  5.7 MiB  1017 MiB  0.65  0.60    0      up  # <--------------- PGS == 0
 3    ssd  0.00099   1.00000  1 GiB   15 MiB  1.6 MiB      0 B   13 MiB  1009 MiB  1.45  1.34   89      up
 0    ssd  0.00099   1.00000  1 GiB   15 MiB  1.1 MiB      0 B   14 MiB  1009 MiB  1.44  1.33   44      up
 2    ssd  0.00099   1.00000  1 GiB   10 MiB  1.4 MiB      0 B  8.6 MiB  1014 MiB  0.98  0.91   45      up
                       TOTAL  6 GiB   66 MiB  7.5 MiB  1.7 KiB   59 MiB   5.9 GiB  1.08
MIN/MAX VAR: 0.60/1.34  STDDEV: 0.28
```

I cordoned the node and deleted the OSD pod again:
```
$ kubectl cordon minikube-m02
$ kubectl delete pod -n rook-ceph rook-ceph-osd-1-dfb8dfc98-xxsp4
```

Then, the blocking PDBs were successfully created again (because my PR is applied):
```
$ kubectl -n rook-ceph logs rook-ceph-operator-6644798f58-bj8t4 -f
2024-02-02 01:11:20.170147 I | clusterdisruption-controller: osd "rook-ceph-osd-1" is down and a possible node drain is detected
2024-02-02 01:11:20.615101 I | clusterdisruption-controller: osd is down in failure domain "minikube-m02". pg health: "all PGs in cluster are clean"
2024-02-02 01:11:22.284086 I | clusterdisruption-controller: creating temporary blocking pdb "rook-ceph-osd-host-minikube" with maxUnavailable=0 for "host" failure domain "minikube"
2024-02-02 01:11:22.292043 I | clusterdisruption-controller: creating temporary blocking pdb "rook-ceph-osd-host-minikube-m03" with maxUnavailable=0 for "host" failure domain "minikube-m03"
2024-02-02 01:11:22.296280 I | clusterdisruption-controller: deleting the default pdb "rook-ceph-osd" with maxUnavailable=1 for all osd

$ kubectl get pdb -n rook-ceph ### <----------- The PDBs are correctly set.
NAME                              MIN AVAILABLE   MAX UNAVAILABLE   ALLOWED DISRUPTIONS   AGE
rook-ceph-osd-host-minikube       N/A             0                 0                     29s
rook-ceph-osd-host-minikube-m03   N/A             0                 0                     29s
```

Reply:
@sp98 I re-confirmed that my PR works. I uploaded the scripts I used here: https://github.com/ushitora-anqou/rook-on-minikube/tree/rook-13511

Here's the result:

I deployed a Rook/Ceph cluster using my PR with 3 nodes and 6 OSDs, and also deployed a RGW.
```
$ kubectl exec -n rook-ceph -it deploy/rook-ceph-tools -- ceph osd tree
ID  CLASS  WEIGHT   TYPE NAME              STATUS  REWEIGHT  PRI-AFF
-1         0.00595  root default                                    
-7         0.00198      host minikube                               
 4    ssd  0.00099          osd.4              up   1.00000  1.00000
 5    ssd  0.00099          osd.5              up   1.00000  1.00000
-5         0.00198      host minikube-m02                           
 1    ssd  0.00099          osd.1              up   1.00000  1.00000
 3    ssd  0.00099          osd.3              up   1.00000  1.00000
-3         0.00198      host minikube-m03                           
 0    ssd  0.00099          osd.0              up   1.00000  1.00000
 2    ssd  0.00099          osd.2              up   1.00000  1.00000

$ kubectl exec -n rook-ceph -it deploy/rook-ceph-tools -- ceph osd df
ID  CLASS  WEIGHT   REWEIGHT  SIZE   RAW USE  DATA     OMAP  META     AVAIL     %USE  VAR   PGS  STATUS
 4    ssd  0.00099   1.00000  1 GiB  9.2 MiB  868 KiB   0 B  8.3 MiB  1015 MiB  0.90  0.83   46      up
 5    ssd  0.00099   1.00000  1 GiB  9.8 MiB  1.3 MiB   0 B  8.6 MiB  1014 MiB  0.96  0.90   43      up
 1    ssd  0.00099   1.00000  1 GiB   10 MiB  1.2 MiB   0 B  8.8 MiB  1014 MiB  0.98  0.91   31      up
 3    ssd  0.00099   1.00000  1 GiB   13 MiB  904 KiB   0 B   13 MiB  1011 MiB  1.31  1.22   58      up
 0    ssd  0.00099   1.00000  1 GiB   14 MiB  932 KiB   0 B   13 MiB  1010 MiB  1.38  1.28   50      up
 2    ssd  0.00099   1.00000  1 GiB  9.3 MiB  1.2 MiB   0 B  8.1 MiB  1015 MiB  0.91  0.85   39      up
                       TOTAL  6 GiB   66 MiB  6.4 MiB   0 B   60 MiB   5.9 GiB  1.07                   
MIN/MAX VAR: 0.83/1.28  STDDEV: 0.20
```

The default PDB was correctly set:
```
$ kubectl get pdb -n rook-ceph
NAME            MIN AVAILABLE   MAX UNAVAILABLE   ALLOWED DISRUPTIONS   AGE
rook-ceph-osd   N/A             1                 1                     106s
```

Then, I cordoned a node and deleted a OSD pod on the node to emulate draining of the node:
```
$ kubectl cordon minikube-m02
node/minikube-m02 cordoned

$ kubectl delete pod -n rook-ceph rook-ceph-osd-1-dfb8dfc98-chmbj
pod "rook-ceph-osd-1-dfb8dfc98-chmbj" deleted
```

Then, the blocking PDBs were successfully created (as before my PR is applied):
```
$ kubectl logs -n rook-ceph rook-ceph-operator-6644798f58-bj8t4 -f
2024-02-02 01:02:57.355897 I | clusterdisruption-controller: osd "rook-ceph-osd-1" is down and a possible node drain is detected
2024-02-02 01:02:57.788077 I | clusterdisruption-controller: osd is down in failure domain "minikube-m02". pg health: "cluster is not fully clean. PGs: [{StateName:active+clean Count:57} {StateName:active+undersized Count:22} {StateName:active+undersized+degraded Count:10}]"
2024-02-02 01:02:59.094658 I | clusterdisruption-controller: creating temporary blocking pdb "rook-ceph-osd-host-minikube" with maxUnavailable=0 for "host" failure domain "minikube"
2024-02-02 01:02:59.102011 I | clusterdisruption-controller: creating temporary blocking pdb "rook-ceph-osd-host-minikube-m03" with maxUnavailable=0 for "host" failure domain "minikube-m03"
2024-02-02 01:02:59.107224 I | clusterdisruption-controller: deleting the default pdb "rook-ceph-osd" with maxUnavailable=1 for all osd

$ kubectl get pdb -n rook-ceph ### <------------ The PDBs are correctly set
NAME                              MIN AVAILABLE   MAX UNAVAILABLE   ALLOWED DISRUPTIONS   AGE
rook-ceph-osd-host-minikube       N/A             0                 0                     2m14s
rook-ceph-osd-host-minikube-m03   N/A             0                 0                     2m14s
```

Next, I uncordoned the node, and made the OSD have no PGs:
```
$ kubectl uncordon minikube-m02

$ kubectl get pdb -n rook-ceph
NAME            MIN AVAILABLE   MAX UNAVAILABLE   ALLOWED DISRUPTIONS   AGE
rook-ceph-osd   N/A             1                 1                     5s

$ kubectl exec -n rook-ceph deploy/rook-ceph-tools -- ceph osd crush reweight osd.1 0
reweighted item id 1 name 'osd.1' to 0 in crush map

$ kubectl exec -n rook-ceph deploy/rook-ceph-tools -- ceph osd df
ID  CLASS  WEIGHT   REWEIGHT  SIZE   RAW USE  DATA     OMAP     META     AVAIL     %USE  VAR   PGS  STATUS
 4    ssd  0.00099   1.00000  1 GiB  9.7 MiB  1.0 MiB      0 B  8.7 MiB  1014 MiB  0.95  0.88   49      up
 5    ssd  0.00099   1.00000  1 GiB   10 MiB  1.4 MiB      0 B  8.9 MiB  1014 MiB  1.01  0.93   40      up
 1    ssd        0   1.00000  1 GiB  6.7 MiB  956 KiB    1 KiB  5.7 MiB  1017 MiB  0.65  0.60    0      up  # <--------------- PGS == 0
 3    ssd  0.00099   1.00000  1 GiB   15 MiB  1.6 MiB      0 B   13 MiB  1009 MiB  1.45  1.34   89      up
 0    ssd  0.00099   1.00000  1 GiB   15 MiB  1.1 MiB      0 B   14 MiB  1009 MiB  1.44  1.33   44      up
 2    ssd  0.00099   1.00000  1 GiB   10 MiB  1.4 MiB      0 B  8.6 MiB  1014 MiB  0.98  0.91   45      up
                       TOTAL  6 GiB   66 MiB  7.5 MiB  1.7 KiB   59 MiB   5.9 GiB  1.08
MIN/MAX VAR: 0.60/1.34  STDDEV: 0.28
```

I cordoned the node and deleted the OSD pod again:
```
$ kubectl cordon minikube-m02
$ kubectl delete pod -n rook-ceph rook-ceph-osd-1-dfb8dfc98-xxsp4
```

Then, the blocking PDBs were successfully created again (because my PR is applied):
```
$ kubectl -n rook-ceph logs rook-ceph-operator-6644798f58-bj8t4 -f
2024-02-02 01:11:20.170147 I | clusterdisruption-controller: osd "rook-ceph-osd-1" is down and a possible node drain is detected
2024-02-02 01:11:20.615101 I | clusterdisruption-controller: osd is down in failure domain "minikube-m02". pg health: "all PGs in cluster are clean"
2024-02-02 01:11:22.284086 I | clusterdisruption-controller: creating temporary blocking pdb "rook-ceph-osd-host-minikube" with maxUnavailable=0 for "host" failure domain "minikube"
2024-02-02 01:11:22.292043 I | clusterdisruption-controller: creating temporary blocking pdb "rook-ceph-osd-host-minikube-m03" with maxUnavailable=0 for "host" failure domain "minikube-m03"
2024-02-02 01:11:22.296280 I | clusterdisruption-controller: deleting the default pdb "rook-ceph-osd" with maxUnavailable=1 for all osd

$ kubectl get pdb -n rook-ceph ### <----------- The PDBs are correctly set.
NAME                              MIN AVAILABLE   MAX UNAVAILABLE   ALLOWED DISRUPTIONS   AGE
rook-ceph-osd-host-minikube       N/A             0                 0                     29s
rook-ceph-osd-host-minikube-m03   N/A             0                 0                     29s
```

Reply:
@ushitora-anqou Thanks for the details. Can you please confirm the pdb behavior for the following scenarios as well?
- This disk backing the OSD was removed.
- OSD deployment was deleted.

Reply:
@sp98 The blocking PDBs were successfully created in the both scenarios. The details are: https://github.com/ushitora-anqou/rook-on-minikube/tree/4332b90d438c2a919bef3e9bef8f4041795e913d?tab=readme-ov-file#additional-investigation

Reply:
Thanks for the detailed testing @ushitora-anqou . Really appreciate the effort you have put into testing this. 

@travisn The PR looks good to me to I'll approve this. I don't think there will be any regression due to this change (fingers crossed).  But still requesting your approval before we merge this, just be double sure that I'm not missing anything.

Reply:
I understand the changes in this PR, but I want to be clear about the behavior change. If any OSD pod is down, now we would expect the blocking PDBs to be in place. Even if the PGs become active+clean, the blocking PDBs will always be enabled. This means that if an OSD is down due to a failed disk, that OSD would either have to be repaired or purged before the PDBs will be reset again. 

This behavior is simple and intuitive to me, but it is different from the previous behavior that would allow the PDBs to be reset after some time even if an OSD disk fails and Ceph backfills its data to other OSDs.

Reply:
It seems correct to me.

> it is different from the previous behavior that would allow the PDBs to be reset after some time even if an OSD disk fails and Ceph backfills its data to other OSDs.

Yes, it is. And my PR claims that the previous behavior was problematic because the reset PDBs didn't allow the OSDs that could be drained to actually be drained.